src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs

```diff
@@ -268,6 +268,31 @@ private static NavigateToMatchKind GetNavigateToMatchKind(ArrayBuilder<PatternMa
                 return NavigateToMatchKind.Substring;
             }
 
+            if (nameMatches.Any(r => r.Kind == PatternMatchKind.CamelCaseExact))
```

Comment:
This method won't give the right match kinds. Consider the case of trying to discover NavigateToMatchKind. If the user searches "Navigate Kind", you'll have two name matches, one with Prefix, and one with Substring. Here, you'll return Prefix, but the correct kind is CamelCaseNonContiguousPrefix. I hit the same issue internally to VS, and wrote the comment below for how we're merging any two match kinds.

```
            // Guess the match kind. Unfortunately we can't assume contiguity. It is hard to guess when dealing with Exact, Fuzzy,
            // and multiple prefixes, so assume that if you have an exact match, the other merged thing must be redundant.
            // The only exception is fuzzy (which degrades everything) since the match obviously missed somehow.
            //
            // Kinds:
            // Exact (E), Prefix (P), Substring (S), CamelCaseExact (CE), CamelCasePrefix (CP), CamelCaseNonContiguousPrefix (NP)
            // CamelCaseSubstring (CS), CamelCaseNonContiguousSubstring(NS ), Fuzzy (F)
            //
            // Mapping:
            // In |E  |P  |S  |CE |CP |NP |CS |NS |F
            //     ___________________________________
            // E  |E   E   E   E   E   E   E   E   F
            // P  |    P   NP  CE  CP  NP  NP  NP  F
            // S  |        NS  CE  NP  NP  NS  NS  F
            // CE |            CE  CE  CE  CE  CE  F
            // CP |                CP  NP  NP  NP  F
            // NP |                    NP  NP  NP  F
            // CS |                        NS  NS  F
            // NS |                            NS  F
            // F  |                                F
```

Reply:
If you have access to our sources, take a look at the full algorithm here: https://devdiv.visualstudio.com/_search?type=code&lp=apps-projects-Collection&text=patternmatchextensions.cs&result=DefaultCollection%2FDevDiv%2FVS-Platform%2FGBmaster%2F%2Fsrc%2FEditor%2FText%2FImpl%2FPatternMatching%2FPatternMatchExtensions.cs&preview=1&_a=contents

Reply:
I'm not convinced there is a 'correct' mapping here.  Basically, if the user has space separated patterns, we can match them all in very different ways.  even if they were all 'Exact' that doesn't mean that the final result should be 'Exact'.  But we're effectively trying to say "as long as at least one of these was at this level, consider that the final value". 

> But the correct kind is CamelCaseNonContiguousPrefix

that's the problem though.  That's may not be the correct kind.   :(

Reply:
I'm not sure how we'd get to your case of having several exact matches without them all having the same characters. Regardless, I'm mostly saying that if you get two results and one is Exact and the other is Fuzzy, returning exact is probably not the right thing to do. The table I came up with above was my best effort to say, given two space-separated search strings, and their respective match kinds, what is a decent guess as to the kind of match we actually got.

Reply:
> I'm mostly saying that if you get two results and one is Exact and the other is Fuzzy, returning exact is probably not the right thing to do.

I'm not convinced that that's the case.  Indeed, i'm fairly certain we actually shifted to that to deal with some cases where returning non-exact ended up deprioritizing the result over results that it was better than. 

One way we could tell would be to introduce your table and see what changes in our tests.  Then evaluate if that's a problem or not. 

If it's fine, then great!   if things regress, we can look into how to deal with that.

Reply:
well, I suggest first finishing implementing current approach and then check what would be an impact of introducing this kind of merging on Roslyn tests.

Reply:
@grabalon can you please help @chborl with checking this?

Reply:
I opened a new issue to track this https://github.com/dotnet/roslyn/issues/24294

Reply:
I went through and flagged a few odd cases in our tests that looked like they might have gotten the wrong one which I think are cases we're trying to fix? If it's from this algorithm then we'll need to fix it up here.

Reply:
i don't think it's this algorithm.  I think the issue is that we don't handle _ well.